src/main/kotlin/uk/gov/justice/digital/hmpps/hmppschallengesupportinterventionplanapi/service/CsipRecordService.kt

```diff
@@ -0,0 +1,69 @@
+import uk.gov.justice.digital.hmpps.hmppschallengesupportinterventionplanapi.enumeration.ReferenceDataType.INCIDENT_LOCATION
+import uk.gov.justice.digital.hmpps.hmppschallengesupportinterventionplanapi.enumeration.ReferenceDataType.INCIDENT_TYPE
+import uk.gov.justice.digital.hmpps.hmppschallengesupportinterventionplanapi.model.CsipRecord
+import uk.gov.justice.digital.hmpps.hmppschallengesupportinterventionplanapi.model.request.CreateContributoryFactorRequest
+import uk.gov.justice.digital.hmpps.hmppschallengesupportinterventionplanapi.model.request.CreateCsipRecordRequest
+import uk.gov.justice.digital.hmpps.hmppschallengesupportinterventionplanapi.repository.CsipRecordRepository
+import uk.gov.justice.digital.hmpps.hmppschallengesupportinterventionplanapi.repository.ReferenceDataRepository
+
+@Service
+class CsipRecordService(
+  val referenceDataRepository: ReferenceDataRepository,
+  val csipRecordRepository: CsipRecordRepository,
+  val prisonerSearchClient: PrisonerSearchClient,
+) {
+
+  @Transactional
+  fun createCsipRecord(
+    request: CreateCsipRecordRequest,
+    prisonNumber: String,
+    requestContext: CsipRequestContext,
+  ): CsipRecord {
+    val prisoner = prisonerSearchClient.getPrisoner(prisonNumber)
+    require(prisoner != null) { "Prisoner with prison number $prisonNumber could not be found" }
+    val incidentType = getReferenceData(INCIDENT_TYPE, request.referral.incidentTypeCode)
+    val incidentLocation = getReferenceData(INCIDENT_LOCATION, request.referral.incidentLocationCode)
+    val referrerAreaOfWork = getReferenceData(AREA_OF_WORK, request.referral.refererAreaCode)
+    val incidentInvolvement = getReferenceData(INCIDENT_INVOLVEMENT, request.referral.incidentInvolvementCode)
+
+    val contributoryFactors: Map<String, ReferenceData> = request.referral.contributoryFactors.let { factors ->
+      val factorTypes = referenceDataRepository.findByDomain(CONTRIBUTORY_FACTOR_TYPE)
+      factors.associateBy({ it.factorTypeCode }, { it.getFactorType(factorTypes) })
+    }
+
+    val record = request.toCsipRecord(prisonNumber, requestContext).create(
+      createCsipRecordRequest = request,
+      csipRequestContext = requestContext,
+      incidentType = incidentType,
+      incidentLocation = incidentLocation,
+      referrerAreaOfWork = referrerAreaOfWork,
+      incidentInvolvement = incidentInvolvement,
+      contributoryFactors = contributoryFactors,
+    )
+    return csipRecordRepository.saveAndFlush(record).toModel()
+  }
+
+  private fun getReferenceData(type: ReferenceDataType, code: String) =
+    referenceDataRepository.findByDomainAndCode(type, code)?.also {
+      require(it.isActive()) { "$type code '$code' is inactive" }
+    } ?: throw IllegalArgumentException("$type code '$code' does not exist")
+
+  private fun CreateContributoryFactorRequest.getFactorType(roles: Collection<ReferenceData>): ReferenceData {
+    return roles.find { it.code.equals(factorTypeCode) }?.also {
```

Comment:
This will throw an exception on the first missing or inactive code. Ideally missing or inactive codes should be concatenated to a single error message e.g. “Contributory factor code(s) ‘AAA’, ‘BBB’ not found or are inactive”.

This was a late update to the ticket though so feel free to ignore.